File: scripts/utils/report_utils.js
const path = require("path");
const { get_deployment_version } = require("./dbt_utils");
const {
  read_file,
  write_file,
  ensure_directory_exists,
} = require("./file_utils");

const SCHEMA = "reporting";
const ROLE = "reporting";
const BASE_DIR = path.resolve(__dirname, "../..");
const REPORTS_DIR = path.resolve(BASE_DIR, "compiled", "reports");
const VIEWS_DIR = path.resolve(BASE_DIR, "compiled", "views");

/**
 * Compiles a report by reading a SQL file and its associated configuration,
 * modifying the configuration, and then writing the result to an output file.
 *
 * @param {string} database - The database name used to replace in the SQL query.
 * @param {string} sql_file - The path to the SQL file to read.
 * @param {string} config_file - The path to the JSON configuration file.
 * @param {string} output_file - The path where the compiled report will be written.
 */
function compile_report(database, sql_file, config_file, output_file) {
  try {
    const sql = read_file(sql_file);
    const json = read_file(config_file, "json");

    json.query = sql
      .replace(/\r?\n\s+/g, "\n")
      .replace(new RegExp(`"${database}".`, "g"), "");

    json.db_schema = SCHEMA;
    write_file(output_file, json, "json");
  } catch (err) {
    console.error(`Error processing files: ${err.message}`);
    process.exit(1);
  }
}

/**
 * Generates project reports based on a target tag and a manifest file.
 * It filters models in the manifest, compiles reports for matching models,
 * and writes them to a specific directory.
 *
 * @param {string} target - The target tag used to filter models.
 */
function generate_project_reports(target) {
  const manifest = read_file(
    path.join(BASE_DIR, "target", "manifest.json"),
    "json"
  );
  const nodes = Object.keys(manifest.nodes).filter(
    (key) =>
      key.startsWith("model") &&
      manifest.nodes[key].tags &&
      manifest.nodes[key].tags.includes("reports") &&
      manifest.nodes[key].tags.includes(target)
  );

  if (nodes.length === 0) {
    console.warn(`No report models found for target: ${target}`);
    return;
  }

  ensure_directory_exists(REPORTS_DIR);

  for (const node of nodes) {
    const report = manifest.nodes[node];
    const sql_file = path.join(BASE_DIR, report.compiled_path);
    const config_file = path
      .join(BASE_DIR, report.original_file_path)
      .replace(".sql", ".json")
      .replace("sql", "config");
    const output_file = path.join(REPORTS_DIR, `${report.name}.json`);

    compile_report(report.database, sql_file, config_file, output_file);
    console.log(`Compiled report: ${report.name}.sql`);
  }
}

/**
 * Generates the import reports script to process the compiled reports.
 * This script will execute the import commands for each report JSON file.
 */
function generate_import_report_script() {
  const script_content = `const fs = require("fs");
  const path = require("path");
  const { exec } = require("child_process");
  const { readFile } = require("fs/promises");

  const folderPath = path.resolve(".");
  const baseCommand = "node ./dist/app.bundle.js importReport";

  fs.readdir(folderPath, async (err, files) => {
    if (err) {
      console.error(\`Error reading directory: \${err.message}\`);
      return;
    }

    const jsonFiles = files.filter((file) => file.endsWith(".json"));

    if (jsonFiles.length === 0) {
      console.log("No JSON files found in the folder.");
      return;
    }

    for (const file of jsonFiles) {
      const filePath = path.join(folderPath, file);

      try {
        const fileContent = await readFile(filePath, "utf-8");
        const json = JSON.parse(fileContent);
        const reportName = json.name;

        const command = \`\${baseCommand} -n '\${reportName}' -f "\${filePath}"\`;

        console.log(\`Executing command: \${command}\`);

        exec(command, (err, stdout, stderr) => {
          if (err) {
            console.error(\`Error executing command for file \${file}: \${err.message}\`);
            return;
          }

          if (stderr) {
            console.error(\`Error output for file \${file}: \${stderr}\`);
            return;
          }

          console.log(\`Success for file \${file}: \${stdout}\`);
        });
      } catch (err) {
        console.error(\`Error processing file \${file}: \${err.message}\`);
      }
    }
  });`;

  ensure_directory_exists(REPORTS_DIR);

  const output_path = path.join(REPORTS_DIR, "import_reports.js");

  try {
    write_file(output_path, script_content);
    console.log(`Script created successfully at: ${output_path}`);
  } catch (err) {
    console.error(`Failed to write the script file: ${err.message}`);
  }
}

/**
 * Generates a reporting schema script by compiling models in the targeted project.
 * It orders models based on their dependencies, and writes a script to create or replace views.
 *
 * @param {string} target - The target tag used to filter models.
 */
function generate_reporting_schema_script(target) {
  const manifest = read_file(
    path.join(BASE_DIR, "target", "manifest.json"),
    "json"
  );
  const nodes = Object.keys(manifest.nodes).filter(
    (key) =>
      key.startsWith("model") &&
      manifest.nodes[key].tags &&
      !manifest.nodes[key].tags.includes("reports")
  );

  if (nodes.length === 0) {
    console.warn(`No models found with the target: ${target}`);
    return;
  }

  let processedNodes = new Set();
  let orderedNodes = [];

  while (processedNodes.size < nodes.length) {
    let currentLevel = nodes.filter((node) => {
      let dependencies = manifest.nodes[node].depends_on.nodes || [];
      return dependencies.every(
        (dep) => dep.startsWith("source") || processedNodes.has(dep)
      );
    });

    if (currentLevel.length === 0) {
      console.error(
        "Error: Circular dependency detected or missing dependencies."
      );
      process.exit(1);
    }

    currentLevel.forEach((node) => {
      processedNodes.add(node);
      orderedNodes.push(node);
    });
  }

  const scripts = [
    `drop schema if exists ${SCHEMA} cascade;`,
    `create schema ${SCHEMA};`,
    `alter default privileges in schema ${SCHEMA} grant select on tables to ${ROLE};`,
  ];

  orderedNodes.forEach((node) => {
    const model = manifest.nodes[node];
    const compiled_path = path.join(BASE_DIR, model.compiled_path);
    const sql = read_file(compiled_path).replace(
      new RegExp(`"${model.database}"\\.`, "g"),
      ``
    );

    scripts.push(
      `create or replace view "${SCHEMA}"."${model.name}" as (\n${sql}\n);`
    );
  });

  ensure_directory_exists(VIEWS_DIR);
  write_file(
    path.join(
      VIEWS_DIR,
      `reporting_schema_build_script_v${get_deployment_version()}.sql`
    ),
    scripts.join("\n")
  );
}

module.exports = {
  compile_report,
  generate_project_reports,
  generate_import_report_script,
  generate_reporting_schema_script,
};

File: scripts/utils/report_utils.py
import json
import os
import re

from .dbt_utils import (
    get_dbt_project_vars,
    get_deployment_name,
    get_deployment_version,
    get_project_name,
)
from .file_utils import ensure_directory_exists, read_file, write_file
from .system_utils import cprint, execute_command

SCHEMA = "reporting"
ROLE = "tamanu_reporting"
BASE_DIR = os.getcwd()
PROJECT_NAME = get_project_name()
DEPLOYMENT = get_deployment_name()
VERSION = get_deployment_version()
DBT_PACKAGE_DIR = os.path.join(BASE_DIR, "dbt_packages", "tamanu_source_dbt")
VERSION_DIR = os.path.join(BASE_DIR, "compiled", f"v{VERSION}")


def compile_report(database, sql_file, config_file, output_file):
    """
    Compiles a report by processing the SQL and config files and generating a JSON output.

    Args:
        database (str): The name of the database used in the report.
        sql_file (str): The path to the SQL file.
        config_file (str): The path to the configuration file.
        output_file (str): The path to the output JSON file.

    Returns:
        None: Writes the compiled configuration to the output file.
    """
    try:
        sql = read_file(sql_file)
        config = read_file(config_file, "json")

        query = re.sub(r"\r?\n\s+", "\n", sql)
        config["query"] = re.sub(f'"{database}"\\.', "", query)
        version = ".".join(VERSION.split(".")[:2])
        config["tamanuVersion"] = f"~{version}.0"

        write_file(output_file, config, "json")
    except Exception as e:
        cprint(f"Error processing files: {e}", "error")
        exit(1)


def generate_project_reports(language):
    """
    Generates reports for the given target by compiling model nodes tagged with "reports".
    Nodes tagged "restricted" (i.e. sensitive-facility dataset views) are excluded when
    has_sensitive_facility is false. has_sensitive_facility is read from dbt_project.yml via get_dbt_project_vars(),
    not from the dbt runtime context.

    Args:
        language (str): The language to use for report generation.

    Returns:
        None: Creates compiled report JSON files in the reports directory.
    """
    manifest_path = os.path.join(BASE_DIR, "target", "manifest.json")
    manifest = read_file(manifest_path, "json")

    project_vars = get_dbt_project_vars()
    has_sensitive_facility = project_vars.get("has_sensitive_facility", False)

    nodes = [
        key
        for key in manifest["nodes"]
        if key.startswith("model")
        and "reports" in manifest["nodes"][key].get("tags", [])
        and (
            has_sensitive_facility
            or "restricted" not in manifest["nodes"][key].get("tags", [])
        )
    ]

    if not nodes:
        cprint(f"No report models found", "error")
        return

    ensure_directory_exists(VERSION_DIR)

    for node in nodes:
        report = manifest["nodes"][node]

        sql_file = os.path.join(BASE_DIR, report["compiled_path"])
        config_file = (
            os.path.join(
                (
                    DBT_PACKAGE_DIR
                    if report["package_name"] != PROJECT_NAME
                    else BASE_DIR
                ),
                report["original_file_path"],
            )
            .replace(".sql", ".json")
            .replace("sql", "config")
        )

        # Include language in filename (only add suffix if not default)
        lang_suffix = f"-{language}" if language != "default" else ""
        filename = f"{report['name']}-v{VERSION}-{DEPLOYMENT}{lang_suffix}.json"
        output_file = os.path.join(VERSION_DIR, filename)

        compile_report(report["database"], sql_file, config_file, output_file)
        cprint(f"Compiled report: {filename}", "success")


def generate_import_report_script():
    """
    Generates a Python script to import the compiled report JSON files.

    The generated script will iterate over all JSON files in the reports directory
    and execute a command to import each report using a Node.js application.

    Returns:
        None: Writes the Python script to the reports directory.
    """
    script = """const fs = require("fs");
const path = require("path");
const { exec } = require("child_process");

const folderPath = path.resolve(".");
// Build-less images (Tamanu 2.60+) run the CLI from TS source via tsx and ship no dist/
// bundle; older images ship the bundled ./dist/app.bundle.js. Pick whichever is present.
const distBundle = "./dist/app.bundle.js";
const baseCommand = fs.existsSync(distBundle)
  ? `node ${distBundle} importReport`
  : "node --import tsx app importReport";

fs.readdir(folderPath, async (err, files) => {
  if (err) {
    console.error(`Error reading directory: ${err.message}`);
    return;
  }

  const jsonFiles = files.filter((file) => file.endsWith(".json"));
  if (jsonFiles.length === 0) {
    console.log("No JSON files found in the folder.");
    return;
  }

  for (const file of jsonFiles) {
    const filePath = path.join(folderPath, file);
    try {
      const fileContent = await fs.promises.readFile(filePath, "utf-8");
      const json = JSON.parse(fileContent);
      const reportName = json.name;
      const command = `${baseCommand} -n '${reportName}' -f "${filePath}"`;
      console.log(`Executing command: ${command}`);
      exec(command, (err, stdout, stderr) => {
        if (err) {
          console.error(`Error executing command for file ${file}: ${err.message}`);
          return;
        }
        if (stderr) {
          console.error(`Error output for file ${file}: ${stderr}`);
          return;
        }
        console.log(`Success for file ${file}: ${stdout}`);
      });
    } catch (err) {
      console.error(`Error processing file ${file}: ${err.message}`);
    }
  }
});
"""

    ensure_directory_exists(VERSION_DIR)
    output_path = os.path.join(VERSION_DIR, "import_reports.js")
    write_file(output_path, script)

    cprint(f"Script created successfully at: {output_path}", "success")


class ReportingSchemaDependencyError(Exception):
    """A model cannot be ordered into the reporting schema build script."""


def _describe_excluded_dependency(manifest, dep):
    """Say why a dependency is absent from the reporting schema.

    Args:
        manifest (dict): The parsed dbt manifest.
        dep (str): Unique ID of the dependency.

    Returns:
        str: A phrase naming the resource kind or the tag that keeps it out.
    """
    if dep.startswith("seed"):
        return (
            "a seed, and the reporting schema is dropped and rebuilt from views alone, so "
            "seed rows never reach a deployment -- hold them in a map__ model instead"
        )
    if dep.startswith("snapshot"):
        return "a snapshot, which the reporting schema does not create"

    tags = manifest["nodes"].get(dep, {}).get("tags", [])
    if "restricted" in tags:
        return 'tagged "restricted", which is excluded while has_sensitive_facility is false'
    for tag in ("reports", "internal"):
        if tag in tags:
            return f'tagged "{tag}", which is excluded from the reporting schema'
    return "not a model the reporting schema creates"


def _describe_stall(manifest, remaining, selectable):
    """Explain why the ordering stopped with models still unplaced.

    Args:
        manifest (dict): The parsed dbt manifest.
        remaining (list): Unique IDs of the models still unordered.
        selectable (set): Unique IDs of every model the reporting schema will create.

    Returns:
        str: A multi-line explanation naming each blocking model and dependency.
    """
    def name(node):
        return manifest["nodes"][node]["name"]

    blocked = {}
    for node in remaining:
        excluded = sorted(
            dep
            for dep in manifest["nodes"][node]["depends_on"]["nodes"]
            if not dep.startswith("source") and dep not in selectable
        )
        if excluded:
            blocked[node] = excluded

    if not blocked:
        return "Circular dependency between reporting schema models: " + ", ".join(
            sorted(name(node) for node in remaining)
        )

    lines = ["Models the reporting schema cannot create:"]
    for node in sorted(blocked, key=name):
        for dep in blocked[node]:
            lines.append(
                f"  {name(node)} depends on {dep}, which is "
                f"{_describe_excluded_dependency(manifest, dep)}"
            )

    waiting = sorted(set(remaining) - set(blocked), key=name)
    if waiting:
        lines.append("Blocked behind them: " + ", ".join(name(node) for node in waiting))
    return "\n".join(lines)


def order_models_for_schema(manifest, nodes):
    """Order models so each view is created after the views it selects from.

    Sources are left out of the ordering: they resolve to tables the deployment already
    has, under their own schema.

    Args:
        manifest (dict): The parsed dbt manifest.
        nodes (list): Unique IDs of the models the reporting schema will create.

    Returns:
        list: `nodes` in creation order.

    Raises:
        ReportingSchemaDependencyError: A model depends on something the reporting schema
            never creates, or the models form a cycle.
    """
    selectable = set(nodes)
    processed = set()
    ordered = []

    while len(processed) < len(nodes):
        current = [
            node
            for node in nodes
            if node not in processed
            and all(
                dep.startswith("source") or dep in processed
                for dep in manifest["nodes"][node]["depends_on"]["nodes"]
            )
        ]
        if not current:
            remaining = [node for node in nodes if node not in processed]
            raise ReportingSchemaDependencyError(
                _describe_stall(manifest, remaining, selectable)
            )

        for node in current:
            processed.add(node)
            ordered.append(node)

    return ordered


def generate_reporting_schema_script():
    """
    Generates a SQL script to create views in the reporting schema.

    The script is generated based on the model nodes that do not have the "reports" or
    "internal" tag and are compiled in the project. Dependencies between models are resolved
    before generating the views in the schema. Nodes tagged "internal" (e.g. the
    metric_definitions registry) are dbt-package-internal and never materialised into the
    deployable reporting schema. Nodes tagged "restricted" (i.e. sensitive-facility dataset
    views) are excluded when has_sensitive_facility is false. has_sensitive_facility is read
    from dbt_project.yml via get_dbt_project_vars(), not from the dbt runtime context.

    Returns:
        None: Writes the SQL schema build script to the views directory.
    """
    manifest_path = os.path.join(BASE_DIR, "target", "manifest.json")
    manifest = read_file(manifest_path, "json")

    project_vars = get_dbt_project_vars()
    has_sensitive_facility = project_vars.get("has_sensitive_facility", False)

    nodes = [
        key
        for key in manifest["nodes"]
        if key.startswith("model")
        and "reports" not in manifest["nodes"][key].get("tags", [])
        and "internal" not in manifest["nodes"][key].get("tags", [])
        and (
            has_sensitive_facility
            or "restricted" not in manifest["nodes"][key].get("tags", [])
        )
    ]

    if not nodes:
        cprint(f"No models found", "error")
        return

    ordered = order_models_for_schema(manifest, nodes)

    scripts = [
        f"drop schema if exists {SCHEMA} cascade;",
        f"create schema {SCHEMA};",
        f"grant usage on schema {SCHEMA} to {ROLE};",
        f"alter default privileges in schema {SCHEMA} grant select on tables to {ROLE};",
    ]

    for node in ordered:
        model = manifest["nodes"][node]
        if model["compiled_path"] is None:
            cprint(f"Model {model['name']} has no compiled path, skipping.", "warning")
            continue
        compiled_sql = read_file(os.path.join(BASE_DIR, model["compiled_path"]))
        cleaned_sql = re.sub(f'"{model["database"]}"\\.', "", compiled_sql)
        name = model.get("config", {}).get("alias") or model["name"]
        scripts.append(
            f'create or replace view "{SCHEMA}"."{name}" as (\n{cleaned_sql}\n);'
        )

    ensure_directory_exists(VERSION_DIR)
    output_file = os.path.join(
        VERSION_DIR, f"reporting-schema-v{VERSION}-{DEPLOYMENT}.sql"
    )
    write_file(output_file, "\n".join(scripts))
